components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/device/UserAuthenticationEndpoint.java

```diff
@@ -68,42 +77,109 @@ public void setDeviceAuthService(DeviceAuthService deviceAuthService) {
-            CommonAuthRequestWrapper commonAuthRequestWrapper = new CommonAuthRequestWrapper(request);
-            commonAuthRequestWrapper.setParameter(Constants.CLIENT_ID, clientId);
-            commonAuthRequestWrapper.setParameter(Constants.RESPONSE_TYPE, Constants.RESPONSE_TYPE_DEVICE);
-            commonAuthRequestWrapper.setParameter(Constants.REDIRECTION_URI, deviceFlowDO.getCallbackUri());
-            if (getScope(userCode) != null) {
-                String scope = String.join(Constants.SEPARATED_WITH_SPACE, getScope(userCode));
-                commonAuthRequestWrapper.setParameter(Constants.SCOPE, scope);
+        try {
+            String userCode = request.getParameter(Constants.USER_CODE);
+            // True when input(user_code) is not REQUIRED.
+            if (StringUtils.isBlank(userCode)) {
+                if (log.isDebugEnabled()) {
+                    log.debug("user_code is missing in the request.");
+                }
+                response.sendRedirect(ServiceURLBuilder.create().addPath(Constants.DEVICE_ENDPOINT_PATH)
+                        .addParameter("error", "invalidRequest").build().getAbsolutePublicURL());
+                return null;
             }
-            commonAuthRequestWrapper.setParameter(Constants.NONCE, userCode);
-            return oAuth2AuthzEndpoint.authorize(commonAuthRequestWrapper, response);
-        } else {
-            if (log.isDebugEnabled()) {
-                log.debug("Incorrect user_code: " + userCode);
+            String clientId = deviceAuthService.getClientId(userCode);
+            DeviceFlowDO deviceFlowDODetails =
+                    DeviceFlowPersistenceFactory.getInstance().getDeviceFlowDAO().getDetailsForUserCode(userCode);
+            if (StringUtils.isNotBlank(clientId) && deviceFlowDODetails != null &&
+                    !isExpiredUserCode(deviceFlowDODetails)) {
+                setCallbackURI(clientId);
+                deviceAuthService.setAuthenticationStatus(userCode);
+                CommonAuthRequestWrapper commonAuthRequestWrapper = new CommonAuthRequestWrapper(request);
+                commonAuthRequestWrapper.setParameter(Constants.CLIENT_ID, clientId);
+                commonAuthRequestWrapper.setParameter(Constants.RESPONSE_TYPE, Constants.RESPONSE_TYPE_DEVICE);
+                commonAuthRequestWrapper.setParameter(Constants.REDIRECTION_URI, deviceFlowDO.getCallbackUri());
+                if (getScope(userCode) != null) {
+                    String scope = String.join(Constants.SEPARATED_WITH_SPACE, getScope(userCode));
+                    commonAuthRequestWrapper.setParameter(Constants.SCOPE, scope);
+                }
+                commonAuthRequestWrapper.setParameter(Constants.NONCE, userCode);
+                return oAuth2AuthzEndpoint.authorize(commonAuthRequestWrapper, response);
+            } else {
+                if (log.isDebugEnabled()) {
+                    log.debug("Incorrect user_code: " + userCode);
```

Comment:
```suggestion
                    log.debug("Incorrect user_code");
```

Lets not log the user code as this is a secret

Reply:
fixed